File: AtCoder/AGC026/C.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

long long valid(const string& S, const string& r, const string& b){
    vector<vector<long long>> dp(r.size()+1, vector<long long>(b.size()+1, 0LL));
    dp[0][0] = 1;
    for(int i=0;i<=r.size();i++){
        for(int j=0;j<=b.size();j++){
            if(i < r.size() && r[i] == S[i+j]) dp[i+1][j] += dp[i][j];
            if(j < b.size() && b[j] == S[i+j]) dp[i][j+1] += dp[i][j];
        }
    }
    return dp[r.size()][b.size()];
}

int main(){
    int N; cin >> N;
    string S; cin >> S;
    string T;
    for(int i=2*N-1;i>=N;i--) T += S[i];
    long long res = 0;
    for(int i=0;i<(1<<N);i++){
        string r, b;
        for(int j=0;j<N;j++){
            if(i&(1<<j)) r += S[j];
            else b += S[j];
        }
        res += valid(T, r, b);
    }
    cout << res << endl;
}

File: AtCoder/AGC026/src/bin/c.rs
use proconio::input;
use proconio::marker::Chars;

fn valid(s: &Vec<char>, r: &Vec<char>, b: &Vec<char>) -> i64 {
    let mut dp = vec![vec![0; b.len() + 1]; r.len() + 1];
    dp[0][0] = 1;
    for i in 0..=r.len() {
        for j in 0..=b.len() {
            if i < r.len() && r[i] == s[i + j] {
                dp[i + 1][j] += dp[i][j];
            }
            if j < b.len() && b[j] == s[i + j] {
                dp[i][j + 1] += dp[i][j];
            }
        }
    }
    dp[r.len()][b.len()]
}

fn main() {
    input! { n: usize, s: Chars}
    let t = (0..n).map(|i| s[2 * n - 1 - i]).collect();
    let mut res = 0;
    for i in 0..1 << n {
        let mut r = Vec::new();
        let mut b = Vec::new();
        for j in 0..n {
            if i & (1 << j) != 0 {
                r.push(s[j]);
            } else {
                b.push(s[j])
            };
        }
        res += valid(&t, &r, &b);
    }
    println!("{}", res);
}
